middleware/spy.go

```diff
@@ -0,0 +1,43 @@
+package middleware
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/gol4ng/httpware"
+)
+
+// Spy middleware allows you to do some actions with the http.Request and http.Response.
+//
+// /!\ this middleware should not be wrapped by middlewares that modify the request and/or response.
+// /!\ the request and response provided to the callback func are copies of the actual request and response,
+//     do not use them to modify the request and/or the response.
+//
+// For example, you can use this middleware to log each request and its related response.
+func Spy(callback func(http.Response)) httpware.Middleware {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+			writerInterceptor := NewResponseWriterInterceptor(writer)
+			buf := new(bytes.Buffer)
+			_, err := buf.ReadFrom(req.Body)
+			if err == nil {
```

Comment:
I'm not a fan of that error handling

@instabledesign any thought on what to do when we can't read the request ?